File: TextToFileEncryptionJava/src/main/java/com/mtnguyen/TextToFileEncryption/Encryption.java
package com.mtnguyen.TextToFileEncryption;

import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;

import javax.crypto.Cipher;
import javax.crypto.CipherInputStream;
import javax.crypto.CipherOutputStream;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import com.google.common.base.Strings;

public class Encryption {
	private static final String SaltKey = "s@ltedK3y";	
	private static final String IvKey = "1V3ct0rk3y";
	
	/**
  	 * This method encrypt string to file using Triple DES
  	 * @param originalString String to be encrypted
  	 * @param strOutputFilePath Full file path for the output 
  	 * @param saltKey Salted key
  	 * @param ivKey Initialization vector key
  	 * @author Man Nguyen
	 * @throws Exception 
  	 * @remark Cross platform with C# encryption
  	 */
  	public static void encrypt3DesTextToFile(String originalString, String strOutputFilePath, String saltKey, String ivKey) throws Exception
	{		
		try(FileOutputStream fos = new FileOutputStream(strOutputFilePath)) {
			String key = Strings.isNullOrEmpty(saltKey) ? SaltKey : saltKey; //need to be 24 bytes
	        String iv = Strings.isNullOrEmpty(ivKey) ? IvKey : ivKey ; //need to be 8 bytes
	        	        	        
	        byte[] plainText = originalString.getBytes(); //get byte from message
	        
			//Create triple DES cipher instance
	        Cipher c3des = create3DesCipher(key, iv, Cipher.ENCRYPT_MODE); 
	        //Writing the cipher to the file
	        try(CipherOutputStream cos = new CipherOutputStream(fos, c3des)) {
	        	cos.write(plainText, 0, plainText.length);
	        }			
		} catch (Exception e) {
			e.printStackTrace();
			throw e;
		} 
	}
  	
  	/**
  	 * This method decrypt file to string using Triple DES
  	 * @param strInputFilePath Full file path of the encrypted file
  	 * @param saltKey Salted key
  	 * @param ivKey Initialization vector key
  	 * @return Decrypted string
  	 * @author Man Nguyen
  	 * @throws Exception 
  	 * @remark Cross platform with C# encryption
  	 */
	public static String decrypt3DesFileToText(String strInputFilePath, String saltKey, String ivKey) throws Exception
	{
		try(InputStream fis = new FileInputStream(strInputFilePath)) {			
			String key = Strings.isNullOrEmpty(saltKey) ? SaltKey : saltKey; //need to be 24 bytes
	        String iv = Strings.isNullOrEmpty(ivKey) ? IvKey : ivKey ; //need to be 8 bytes
	        
			//Create triple DES cipher instance
	        Cipher c3des = create3DesCipher(key, iv, Cipher.DECRYPT_MODE); 
						
			try(CipherInputStream cis = new CipherInputStream(fis, c3des)) {		        
		        ByteArrayOutputStream baos = new ByteArrayOutputStream();
		    	byte[] buffer = new byte[fis.available()];
		    	int numRead;	    	
				//Reading and decrypt cipher from the file
		    	while ((numRead = cis.read(buffer)) >= 0) {
		    		baos.write(buffer, 0, numRead);
		    	}
		    	
		    	return new String(baos.toByteArray());
			}
		} catch (Exception e) {
			e.printStackTrace();
			throw e;
		} 
	}
  	
  	/**
     * This is helper method to create Triple DES Cipher
     * @param saltKey Salted key
     * @param ivKey Initialization vector key
     * @param cipherMode Encrypt or decrypt mode
     * @return Cipher to encrypt or decrypt depend on cipher mode
     * @throws Exception
     * @author Man Nguyen
     */
    private static Cipher create3DesCipher(String saltKey, String ivKey, int cipherMode) throws Exception {
    	final byte[] keyBytes = createSha512Bytes(saltKey, 24);
        final byte[] ivBytes = createSha512Bytes(ivKey, 8);
        
        Cipher cipher = Cipher.getInstance("DESede/CBC/PKCS5Padding");
        SecretKeySpec secretKeySpec = new SecretKeySpec(keyBytes, "DESede");
        IvParameterSpec ivParameterSpec = new IvParameterSpec(ivBytes);
        cipher.init(cipherMode, secretKeySpec, ivParameterSpec);        
        return cipher;
    }
    
    /**
     * This is a helper method to create iv bytes
     * @param input Value to transform into SHA 512 byte array
     * @param copyLength The length to copy
     * @return SHA 512 byte array
     * @author Man Nguyen
     */
    private static byte[] createSha512Bytes(String input, int copyLength)
    {
    	byte[] digestIV = getDigestBytes("SHA-512", input); //64 bytes
    	final int bLength = (copyLength > 64) ? 64 : copyLength; 
    	return Arrays.copyOf(digestIV, bLength);    	
    }
    
    /**
     * This is a helper method to create digest bytes.
     * @param algorithm Name of the algorithm
     * @param input Value to transform into SHA 512 byte array
     * @return Byte array of digest key
     * @author Man Nguyen
     */
    private static byte[] getDigestBytes(String algorithm, String input)
    {
    	try {
			return MessageDigest.getInstance(algorithm).digest(input.getBytes("utf-8"));
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}
    	return null;
    }
}

File: TextToFileEncryptionC#/TextToFileEncryption/Encryption.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Security.Cryptography;
using System.Text;

namespace TextToFileEncryption
{
    /// <summary>
    /// Author: Man Nguyen
    /// </summary>
    public class Encryption
    {
        private const string SaltKey = "s@ltedK3y";
        private const string IvKey = "1V3ct0rk3y";

        #region Encrypt String to file and file back to string
        /// <summary>
        /// Encrypt string into a file using Triple DES
        /// </summary>
        /// <param name="originalString">String to be encrypted </param>
        /// <param name="outputFilePath">Full file path for the output</param>
        /// <param name="saltKey">Salted key</param>
        /// <param name="ivKey">Initialization vector key</param>
        /// <remarks>Cross platform with Java encryption</remarks>
        public static void Encrypt3DesTextToFile(string originalString, string outputFilePath, string saltKey, string ivKey)
        {
            try
            {
                var key = string.IsNullOrEmpty(saltKey) ? SaltKey : saltKey;
                var iv = string.IsNullOrEmpty(ivKey) ? IvKey : ivKey;
                
                using(var fsEncrypted = new FileStream(outputFilePath, FileMode.Create, FileAccess.Write))
                using (var cTransform = Create3DesProvider(CryptoTransformType.Encrypt, key, iv))
                {
                    using (var cryptoStream = new CryptoStream(fsEncrypted, cTransform, CryptoStreamMode.Write))
                    {
                        //Read in the input file, and then write out to the output file
                        byte[] bytearrayinput = ASCIIEncoding.UTF8.GetBytes(originalString);
                        cryptoStream.Write(bytearrayinput, 0, bytearrayinput.Length);
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex.ToString());
                throw;
            }
        }

        /// <summary>
        /// Decrypt file to string using Triple DES
        /// </summary>
        /// <param name="inputFilePath">Full file path of the encrypted file</param>
        /// <param name="saltKey">Salted key</param>
        /// <param name="ivKey">Initialization vector key</param>
        /// <returns>Decrypted string</returns>
        /// <remarks>Cross platform with Java decryption</remarks>
        public static string Decrypt3DesFileToText(string inputFilePath, string saltKey, String ivKey)
        {
            try
            {
                var key = string.IsNullOrEmpty(saltKey) ? SaltKey : saltKey;
                var iv = string.IsNullOrEmpty(ivKey) ? IvKey : ivKey;

                //Create a file stream to read the encrypted file back.
                using(var fsread = new FileStream(inputFilePath, FileMode.Open, FileAccess.Read))
                //Create a DES decryptor from the DES instance.
                using (var cTransform = Create3DesProvider(CryptoTransformType.Decrypt, key, iv))
                {
                    //Create crypto stream set to read and do a 
                    //DES decryption transform on incoming bytes.
                    using (var cryptoStream = new CryptoStream(fsread, cTransform, CryptoStreamMode.Read))
                    {
                        using (var sr = new StreamReader(cryptoStream))
                        {
                            return sr.ReadToEnd();
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex.ToString());
                throw;
            }
        }
        #endregion

        #region Private methods
        /// <summary>
        /// Helper method to create 3DES CrytoTransform Interface
        /// </summary>
        /// <param name="cryptoType">Encrypt or decrypt</param>
        /// <param name="saltedKey">Salted key</param>
        /// <param name="ivKey">Initialization vector key</param>
        /// <returns>Triple DES encryptor or decryptor object</returns>
        private static ICryptoTransform Create3DesProvider(CryptoTransformType cryptoType, string saltedKey, string ivKey)
        {
            try
            {
                byte[] keyBytes = CreateSha512Bytes(saltedKey, 24);
                byte[] ivBytes = CreateSha512Bytes(ivKey, 8);

                using (var tripleDes = new TripleDESCryptoServiceProvider() { Mode = CipherMode.CBC, Padding = PaddingMode.PKCS7 })
                {
                    switch (cryptoType)
                    {
                        case CryptoTransformType.Encrypt:
                            return tripleDes.CreateEncryptor(keyBytes, ivBytes);
                        case CryptoTransformType.Decrypt:
                            return tripleDes.CreateDecryptor(keyBytes, ivBytes);
                    }
                }
            }
            catch (Exception ex)
            {
                Debug.WriteLine(ex.ToString());
                throw;
            }

            return null;
        }

        /// <summary>
        /// This is a helper method to create digest bytes up to 64 bytes array
        /// </summary>
        /// <param name="key">Value to transform into SHA 512 byte array</param>
        /// <param name="copyLength">The length to copy</param>
        /// <returns>SHA 512 byte array</returns>
        private static byte[] CreateSha512Bytes(string key, int copyLength)
        {
            using (var sha512 = new SHA512CryptoServiceProvider())
            {
                var digestIv = sha512.ComputeHash(Encoding.UTF8.GetBytes(key));
                var bLength = (copyLength > 64) ? 64 : copyLength;
                var ivBytes = new byte[bLength];
                Array.Copy(digestIv, ivBytes, bLength);
                return ivBytes;
            }
        }
        #endregion
    }
}
